File: src/leetcode/hash_maps_and_sets/1657_determine_if_two_strings_are_close.ts
// LeetCode 75
// https://leetcode.com/problems/determine-if-two-strings-are-close/?envType=study-plan-v2&envId=leetcode-75

// Time complexity: O(n)
// Space complexity: O(1)

const closeStrings = (word1: string, word2: string): boolean => {
  if (word1 === word2) return true;

  if (word1.length !== word2.length) return false;

  const word1Occurrences = word1.split('').reduce<Record<string, number>>((acc, char) => {
    acc[char] = (acc[char] ?? 0) + 1;

    return acc;
  }, {});

  const word2Occurrences = word2.split('').reduce<Record<string, number>>((acc, char) => {
    acc[char] = (acc[char] ?? 0) + 1;

    return acc;
  }, {});

  const word1OccurrencesKeys = Object.keys(word1Occurrences).sort().join('');
  const word2OccurrencesKeys = Object.keys(word2Occurrences).sort().join('');

  const word1OccurrencesValues = Object.values(word1Occurrences)
    .sort((a, b) => a - b)
    .join('');
  const word2OccurrencesValues = Object.values(word2Occurrences)
    .sort((a, b) => a - b)
    .join('');

  return word1OccurrencesKeys === word2OccurrencesKeys && word1OccurrencesValues === word2OccurrencesValues;
};

export default closeStrings;

File: src/leetcode/hash_maps_and_sets/1657_determine_if_two_strings_are_close.rb
# frozen_string_literal: true

# LeetCode 75
# https://leetcode.com/problems/determine-if-two-strings-are-close/?envType=study-plan-v2&envId=leetcode-75

# Time complexity: O(n)
# Space complexity: O(1)

# @param {String} word1
# @param {String} word2
# @return {Boolean}
def close_strings(word1, word2)
  # hash_map_word1 = word1.chars.each_with_object(Hash.new(0)) { |item, acc| acc[item] += 1 }
  # hash_map_word2 = word2.chars.each_with_object(Hash.new(0)) { |item, acc| acc[item] += 1 }

  return true if word1 == word2
  return false if word1.length != word2.length

  word1_occurrences = word1.chars.tally
  word2_occurrences = word2.chars.tally

  word1_occurrences.keys.sort == word2_occurrences.keys.sort &&
    word1_occurrences.values.sort == word2_occurrences.values.sort
end
